Make JsonPropertyAccessor returned type directly a JsonNode or value

pilak opened the pull request:
According to [that thread](https://stackoverflow.com/questions/60953156/problem-of-selector-access-with-the-jsonpropertyaccessor) on stackoverflow, I'm submitting a PR in order to improve the selector readability on JsonPropertyAccessor.

I made this changes because I noticed that using the same expression passing from different accessors such as MapAccessor, ReflectivePropertyAccessor is working but not with JsonNodePropertyAccessor.
Here's an example that shows the breaking or the success of using or not the new JsonNodePropertyAccessor along with the same SpEL expression:
https://github.com/pilak/examples/blob/master/TestJsonAccessing.java

Commit:
Change Json Converter

Commit:
Make JsonPropertyAccessor returned type directly a JsonNode or value

* add `Comparable` contract to the initial wrapper so that expressions
containing selection/projection with JsonNode filtering are now possible
* render directly the value of a ValueNode. SpEL parser can now evaluate
correctly the object value when the expression deals with filtering on
JsonNode values (<, >, ==...)
* the 2 above changes make this possible:
`property.^[name.getTarget().asText() == 'value1'].name` can now be
written as simply as `property.^[name == 'value1'].name`. The expression
is now fully compatible with other PropertyAccessor types
* rename `ToStringFriendlyJsonNode` to `JsonNodeWrapper` as it reflects
more the capability of the new wrapper
* add a GenericConverter to be able to convert the `JsonNodeWrapper`
class into a `JsonNode` (or its derivative classes) directly while
calling `expression.getValue(..., JsonNode.class)`
* add the possibility to access list items in a json-path like index
	=> a negative number will begin from the end of the list <=
* return null value when JsonNode cannot handle the property name
(or index)

artembilan commented:
Thank you for contribution @pilak !

We will take a look into your PR soon.

Meanwhile, please, consider to run `gradlew clean check` task against your changes: looks like there are a lot of Checkstyle violations.

First of all we use **tabs** for indents. Secondly our imports order is like:
```
<module name="ImportOrder">
	<property name="groups" value="java,/^javax?\./,org,org.springframework,*"/>
	<property name="ordered" value="true"/>
	<property name="separated" value="true"/>
	<property name="option" value="top"/>
	<property name="sortStaticImportsAlphabetically" value="true"/>
</module>
```

pilak commented:
Hi Artem,

I made changes related to checkstyle violations and I think it's ok now.
I don't know if everything is well integrated, especially for the bean context integration part. 

Hope my solution will fit.
Regards

artembilan reviewed (changes requested):
Sorry, I don't understand what is going on.

> Json property accessor selector friendly

This is not enough to justify such a breaking change.

Can you give in the commit message as much explanation what you are doing and why as possible?

See recommendations here: https://chris.beams.io/posts/git-commit/

pilak commented on spring-integration-core/src/main/java/org/springframework/integration/json/JsonPropertyAccessor.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2013-2019 the original author or authors.
+ * Copyright 2014-2020 the original author or authors.
```
sorry about that, I thought `gradlew clean checkstyleMain` was enough

unfortunately I have an error while running the task `check`

`Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Unable to find: src/checkstyle/checkstyle-suppressions.xml
`

artembilan replied:
Can you show, please, how do you call it?
Is it from command line like `./gradlew clean check` ?
Or do you use an IDE integration?

pilak replied:
from the wrapper yes in command line

pilak replied:
The compilation was not working on JDK 11, so I used JDK 8 instead

pilak commented:
> Sorry, I don't understand what is going on.
> 
> > Json property accessor selector friendly
> 
> This is not enough to justify such a breaking change.
> 
> Can you give in the commit message as much explanation what you are doing and why as possible?
> 
> See recommendations here: https://chris.beams.io/posts/git-commit/

Sorry, I don't know exactly what to set for the commit title...
Is the title below good for you ?

`JsonPropertyAccessor now renders a JsonNode or the value if at
leaf`

artembilan commented:
> Is the title below good for you ?

Well, the header of the commit message should not exceed 50 symbols.
So, with any effort it is hard to explain everything in such a short sentence.
Your current title is OK.
What I worry about that there is no explanation what is going on in the change.

And here are some questions from me since I review this:

1. Why would one rename a bean and class making so big breaking change?
2. Why do you change a logic in side that `JsonPropertyAccessor` so radically?
3. Why do you use a proxy over there?

All those are hard changes for me and I can't merge whatever I don't understand. 
With some explanation in the commit message that would be great improvement in our collaboration over this PR.

(Did you really read that Chris Beam's article?)

Thanks

artembilan commented:
@pilak ,

please, take a look here: https://github.com/spring-projects/spring-integration/commit/f047f0528bdfa11731f2e8d0c73ff1aa7d374f68

I think this one has to fix your problem with Checkstyle + Gradle.

artembilan commented:
In general you need to pull `upstream/master` and rebase your `json-property-accessor-selector-friendly` branch to get those changes immediately.

pilak commented:
@artembilan why do we return `"null"` stringified in the link below? Does a `null` not enough?
https://github.com/spring-projects/spring-integration/pull/3270/commits/dea81e3fcf468ea611d9ea8bcd8af7a187b48b63#diff-4c3528130fa3a7cdcb0a2c05f990cd57L209
On the other hand, when the `JsonPropertyAccessor` reads the json object, and then creates a ToStringFriendlyJsonNode, it does check if it is null before and thus returns a TypedValue.NULL if it is. So is that really necessary to override the toString method in ToStringFriendlyJsonNode? I don't get it.

I also found that actually, the _Jackson_ **ArrayNode** should implement the List interface or else Collection. But because not of this, we must `wrap()` the **ArrayNode** before evaluation so that accessing with index can be possible, which should not be the responsability of the caller.
As a workaround we could add the method `wrap(@Nullable Object target)` which defaultly returns the target to the `PropertyAccessor` and call this in the Ast Indexer with `if canRead then wrap() then getValueRef()`

Reading the code again I now understand that this `JsonPropertyAccessor` was primarily designed for the `MessageConverter` purpose. But does it mean we could forget other usages? On different projects on which I work, we use the SpEL expression evaluation as a tool appart, exactly the way I use it in the stackoverflow thread, by declaring all thing manually, and the fact is that everything works, until we use the JsonNode object, so that we had to convert it in other Objects.

Anyway, I will push an update of my changes soon if you still wish, according to your different reviews, and eventually what could be decide from this post.

Thanks for your patience
Regards

pilak commented:
@artembilan I pushed some changes, and modified the commit message to explain better what I intent to solve with my changes
I also changed the first message in that thread
https://github.com/spring-projects/spring-integration/pull/3270#issue-412771865

I will soon answer all of the remaining questions. 
Hope this is not too much time consuming for you, and will worth it.

Regards

artembilan commented:
Thank you, @pilak , for such an effort!
I see now your point, but I'll review it a bit later this or next week.
We have GA release tomorrow and looks like your change is slightly a breaking (or behavior) change. We can't apply it in GA phase.
The next version should be OK.

Stay tuned!

pilak commented:
Ok @artembilan thanks for the feedback.
No problem for the wait, I have all my time. And if the changes are breaking, it is better to have a long-term test phase.
Wish you good luck!

artembilan commented:
Please, rebase your branch to the latest master. It is now `5.4.x`.
We will review your effort shortly after that.
Thanks

pilak commented:
Hi @artembilan,
I rebase from Master and while building on local I have many errros such as:
`org.springframework.amqp.AmqpConnectException: java.net.ConnectException: Connexion refusée (Connection refused)
`
I also have this error from Travis CI
`FluxAggregatorMessageHandlerTests > testWindowTimespan() FAILED
    java.lang.IllegalStateException at FluxAggregatorMessageHandlerTests.java:205`

artembilan commented:
Thank you for the update!

Pulling locally for thorough review...

artembilan commented:
> org.springframework.amqp.AmqpConnectException: java.net.ConnectException: 

I'll turn off my RabbitMQ to see what is wrong in the tests.
Those AMQP tests really have to be covered by the `@RabbitAvailable` condition...

artembilan commented:
It looks like we indeed need a `Comparable` for `JsonNode` and an `ArrayNodeAsList`, but not proxy.
It seems for me a `JsonNode` has all the methods we need need for SpEL collection selection.
Although we need to revise that anyway because it looks like `has()` and `get()` are over there in the `JsonNode`. Probably we don't need to worry about a `List` wrapping anyway. Plus `JsonNode` is an `Iterable` by itself.

I worry that proxying is a heavy operation and we should avoid it as much as possible.

Would you mind to play with much simpler model when we deal only with a `JsonNode`? 
It feels for me that Jackson has made some effort into its core to simplify our life...

Thanks for understanding!

artembilan commented:
> org.springframework.amqp.AmqpConnectException: java.net.ConnectException:

Please, rebase your branch to the latest `master`. The fix should be there for this.

Although consider to test only `spring-integration-core` module: you don't do any changes which may effect other modules.

pilak commented:
> It seems for me a `JsonNode` has all the methods we need need for SpEL collection selection.
> Although we need to revise that anyway because it looks like `has()` and `get()` are over there in the `JsonNode`. Probably we don't need to worry about a `List` wrapping anyway. Plus `JsonNode` is an `Iterable` by itself.

Unfortunately the method below called in AST doesn't check for `Iterable` but instead `Collection`... That's why I guess we used to wrap with `ArrayNodeAsList` at start.
`org.springframework.expression.spel.ast.Indexer.getValueRef(ExpressionState)`

pilak commented:
> It looks like we indeed need a `Comparable` for `JsonNode` and an `ArrayNodeAsList`, but not proxy.
> It seems for me a `JsonNode` has all the methods we need need for SpEL collection selection.
> Although we need to revise that anyway because it looks like `has()` and `get()` are over there in the `JsonNode`.

As JsonNode is an asbtract meta class (in a Jackson way), this is not possible to extend it without having to implement every abstract methods in order to add `Comparable` contract.

artembilan commented:
> this is not possible to extend it without having to implement every abstract methods in order to add Comparable contract. 

OK. Can we deal with the `TreeNode` instead for that `Comparable ` contract? I'm not sure that we indeed should let to have all the `JsonNode` functionality access.
Plus, we probably just could convert an `ArrayNode` into a `List`:
```
 StreamSupport.stream(arrayNode.spliterator(), false)
    .collect(Collectors.toList());
```

pilak commented:
OK why not, it seems to be accessible...
I'll take a look on it asap

artembilan commented:
@pilak ,

gentle ping. What is your current take on the matter?
Thanks

pilak commented:
Hi @artembilan 
Currently I'm very busy. In view of the changes you requested for, this will take a bit of time. I have to adjust some schedules in order to be able to process them.
Regards

artembilan commented:
Sure, @pilak !

Take you time! Your fix is valuable, but we can live without it for a while yet. We have already for a long time, so not a problem to wait for you to be free for this task.

Thank you!

pilak commented:
@artembilan on [that previous comment](https://github.com/spring-projects/spring-integration/pull/3270#issuecomment-626154099), I suggested a workaround to the `wrap()` problem here:

> As a workaround we could add the method wrap(@Nullable Object target) which defaultly returns the target, to the PropertyAccessor interface and then call this in the Ast Indexer with if SUPPORTED_CLASSES...

The target object, is beforehand analyzed by the AST in this method https://github.com/spring-projects/spring-framework/blob/master/spring-expression/src/main/java/org/springframework/expression/spel/ast/Indexer.java#L191, and thus throw a `SpelEvaluationException`, in the case of an ArrayNode.
Do you think we could ask for a PR on `spring-framework` so that the `wrap()` method can be called before any kind of evaluation?

artembilan commented:
@pilak ,

thank you for feedback!

Would you mind to explain more what kind of change you would like to see in AST?
Not fully clear why we can't also go with `wrap()`. My only concern that proxy makes everything a bit slower...

pilak commented:
ok I'll manage the proxy issue first, then back to you with the `wrap()` concern

artembilan commented:
@pilak ,

any update on the matter, please?

We are very close to the `5.4` release this October-November, so would be great to have some solution merged and rejected altogether.

Thanks for understanding!

pilak commented:
Hi @artembilan 
I'll  take back the work soon on the PR, but I'm not sure it will be available for the 5.4 release.
Is that a problem ?
Regards

artembilan commented:
Hi @pilak !

Thank you for you note!

No, it's not a problem. As long as we can make it into the framework one day, it's OK.

I was just curious if you are still available for this task.

Just take your time!

artembilan reviewed (changes requested):
Thank you for the update, @pilak !

I don't think this can make it into the current 5.4 GA phase.
So, probably already to the next one `6.0`.

I'll review it a bit later, next week after release, but meanwhile consider these issue Travis shows to us:
```
/home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/json/JsonPropertyAccessor.java:76: error: cannot find symbol

				return Set.of(new ConvertiblePair(JsonNodeWrapper.class, JsonNode.class));

				          ^

  symbol:   method of(ConvertiblePair)

  location: interface Set

/home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/json/JsonPropertyAccessor.java:123: warning: [cast] redundant cast to JsonNode

			return assertContainerNode((JsonNode) wrapper.getRealNode());

			                           ^

1 error

1 warning
```

Plus see if you rebased to the latest master.

pilak reviewed (commented):
> > this is not possible to extend it without having to implement every abstract methods in order to add Comparable contract.
> 
> OK. Can we deal with the `TreeNode` instead for that `Comparable ` contract? I'm not sure that we indeed should let to have all the `JsonNode` functionality access.
> Plus, we probably just could convert an `ArrayNode` into a `List`:
> 
> ```
>  StreamSupport.stream(arrayNode.spliterator(), false)
>     .collect(Collectors.toList());
> ```

Using this we won't be able to call `expression.getValue(..., ArrayNode.class)`

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/json/JsonPropertyAccessor.java:
```diff
@@ -50,16 +56,32 @@
+		converterRegistry.addConverter(new GenericConverter() {
+			@Override
+			public Set<ConvertiblePair> getConvertibleTypes() {
+				return Collections.singleton(new ConvertiblePair(JsonNodeWrapper.class, JsonNode.class));
```
Why do we need this converter?

If that, shouldn't we introduce it alongside with that `ToStringFriendlyJsonNodeToStringConverter`?
On the other hand: do we still need that `ToStringFriendlyJsonNodeToStringConverter` with this whole your change?

artembilan reviewed (changes requested):
I think it is very close now to the finish.

Would you mind to answer to those my couple questions?
Or make a respective change...

Thanks

artembilan commented:
Please, stop squashing commits. 
I'm not sure now what change have you done since the last review.

pilak reviewed (commented):
> Please, stop squashing commits.
> I'm not sure now what change have you done since the last review.

Ok sorry I will try to restore the history, and make it build better

artembilan reviewed (changes requested):
I think we are good to go with your PR here so far, but now it only can make it into the next `6.0` version since we have just done with `5.4` and such a change cannot make it into the point release since it is slightly breaking change.

Thanks for understanding!

Maxusz666 commented:
:grinning:

artembilan commented:
Since we have decided to go with `5.5` version yet for the upcoming Spring Boot `2.5`, then pulling this change locally for the final review and clean up... I think we can apply such a change into the minor release with an appropriate Migration Guide note (if any).

pilak reviewed (commented):
> Since we have decided to go with `5.5` version yet for the upcoming Spring Boot `2.5`, then pulling this change locally for the final review and clean up... I think we can apply such a change into the minor release with an appropriate Migration Guide note (if any).

Hello @artembilan 
Am I supposed to do something for now?

artembilan commented:
Hi @pilak !
Thank you for reminder!

We've busy with releases this week.
Now we are good to wrap up issues for ucomping `5.5 M1` next week.
So, I'm pulling your PR locally for the final review ASAP.

Probably you don't need to do anything more, unless you'd like to contribute some other stuff 😄 

Anyway I'll update this later.

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/b58b0e5ffaea6f177df4e62d4bb6f9aaf4c82aad.

@pilak ,

thank you very much for the contribution; looking forward for more!

pilak commented:
Nice, thank you @artembilan for your advices
I will probably make some changes then if this PR https://github.com/spring-projects/spring-framework/pull/26323 fits

artembilan commented:
Sure! Looking forward for decision on that PR from Spring Framework commiters!
